hackathon_site/hardware/serializers.py

```diff
@@ -34,9 +38,131 @@ def get_unique_hardware_count(obj: Category):
+            queryset=Hardware.objects.all(), many=False, required=True
+        )
+        quantity = serializers.IntegerField(required=True)
+
+    hardware = OrderCreateHardwareSerializer(many=True, required=True)
+
+    @staticmethod
+    def merge_requests(hardware_requests):
+        # hardware_requests is a list where each element is an OrderedDict
+        # with key value pairs
+        # "id": <Hardware Object>, and
+        # "quantity": <Int>
+        return functools.reduce(
+            lambda x, y: x + y,
+            [Counter({e["id"]: e["quantity"]}) for e in hardware_requests],
+        )
+
+    # check that the requests are within per-team constraints
+    def validate(self, data):
+        # requested_hardware is a Counter where the keys are <Hardware Object>'s and values are <Int>'s
+        requested_hardware = self.merge_requests(hardware_requests=data["hardware"])
+        hardware_query = (
+            Hardware.objects.filter(
+                id__in=[hardware.id for hardware in requested_hardware.keys()],
+            )
+            .all()
+            .prefetch_related("categories", "order_items")
+        )
+        team_unreturned_orders = hardware_query.annotate(
+            past_order_count=Count(
+                "order_items",
+                filter=Q(order_items__part_returned_health__isnull=True)
+                & ~Q(order_items__order__status="Cancelled")
+                & Q(order_items__order__team=self.context["request"].user.profile.team),
+                distinct=True,
+            )
+        )
+        category_counts = dict()
+        error_messages = []
+        for (hardware, requested_quantity) in requested_hardware.items():
+            team_hardware = team_unreturned_orders.get(id=hardware.id)
+            team_hardware_count = getattr(team_hardware, "past_order_count", 0)
+            if hardware.max_per_team < (team_hardware_count + requested_quantity):
+                error_messages.append("Hardware {} limit reached".format(hardware.name))
+            for category in hardware.categories.all():
+                category_counts[category] = (
+                    category_counts.get(category, 0)
+                    + team_hardware_count
+                    + requested_quantity
+                )
+        for (category, count) in category_counts.items():
+            if category.max_per_team < count:
+                error_messages.append("Category {} limit reached".format(category.name))
+        if error_messages:
+            error_message = "; ".join(error_messages)
+            raise serializers.ValidationError(error_message)
+        return data
+
+    def create(self, validated_data):
+        # validated data should already satisfy all constraints
+        requested_hardware = self.merge_requests(
+            hardware_requests=validated_data["hardware"]
+        )
+        new_order = None
+        response_data = {"order_id": -1, "hardware": [], "errors": []}
```

Comment:
so if no new order is created (when all requested quantities or remaining quantities are 0), it would return -1 as order id. idk if thats gonna bother anyone

Reply:
I would say that's not great, since we need to include catches for that on the frontend.

The only way this could happen is if `requested_hardware` is empty, right? I'd say we should either throw a 400 in that case (submitting an order with no items sounds like a bad request to me). 

If you don't think that's appropriate and still want to return 200, then make order_id default to null since that's a bit cleaner to check for on the frontend. But I think 400 (raise a validation error) is appropriate.

Reply:
Ah, it can also end up not making an order if `num_order_items <= 0` for all items in the request (ie, requesting only items that are now sold out). 

Could we do it so that:
- If no hardware was submitted, 400
- If everything they requested was checked out, the order id is None instead of -1. Just set `allow_null=True` on your response serializer I think

Reply:
done :)